### ISSUE TITLE
Splitting #2705 (LiteRT-LM export) into 10 reviewable PRs — review plan and tracking

### ISSUE THREAD

**Issue by pctablet505** [keras-team/keras-hub#2880]
## Complete 28-PR Stacked PR Breakdown (~100–300 LOC per slice)

To ensure every review slice adheres to open-source small PR best practices (~100–300 LOC of incremental changes), we have further partitioned the 10 draft PRs in this tracking issue into **28 Stacked Draft PRs**. Every slice is open as a Draft PR for staged review.

### Architecture & Review Principles
- **Strict Incremental Staging**: Each slice is staged sequentially so that reviewers can inspect small, logically cohesive increments (~100–300 LOC).
- **Caller-less Helper Strategy**: Where an early slice introduces a helper function or spec class before its caller is wired up in a later slice, the PR description explicitly notes the consumer slice in the sequence.
- **Complete Parity**: The cumulative code across each feature sequence is byte-for-byte identical to the original 10-PR branches.

---

### Master Table of Stacked Draft PRs (28 Slices)

| Order | PR Link | Feature Title | ~LOC | What it does & Consumer Mapping |
|---|---|---|---|---|
| **1** | [#2892](https://github.com/keras-team/keras-hub/pull/2892) | **JAX Backend Test Unforcing** | -6 | Stops two test files from forcing JAX backend so tests run multi-backend (#2892) |
| **2** | [#2893](https://github.com/keras-team/keras-hub/pull/2893) | **Model Layer Traceability Fixes** | +22 | Model layer traceability fixes for Llama and SmolLM3 attention (#2893) |
| **3** | [#2908](https://github.com/keras-team/keras-hub/pull/2908) | **Traceable Ops: Core Scope Manager & Basic Shims** | ~310 | Core scope manager (_TraceableOpsScope) and basic shims (one_hot, repeat, amax) |
| **4** | [#2909](https://github.com/keras-team/keras-hub/pull/2909) | **Traceable Ops: Indexing & Slicing Shims** | ~320 | Traceable indexing and slicing shims (slice, take, scatter_update) |
| **5** | [#2894](https://github.com/keras-team/keras-hub/pull/2894) | **Traceable Ops: Attention Shim & Full Parity** | ~232 | Attention shim (dot_product_attention for MHA, GQA, masking) and full op parity |
| **6** | [#2910](https://github.com/keras-team/keras-hub/pull/2910) | **Model Specs: Base Class & Dispatch Registry** | ~340 | LiteRTLMExportSpec base class, dispatch registry, and Llama3/Phi3 specs |
| **7** | [#2911](https://github.com/keras-team/keras-hub/pull/2911) | **Model Specs: Qwen Architecture Family** | ~310 | Qwen architecture family specs (Qwen 2.5, 3, 3.5) and hybrid KV-cache |
| **8** | [#2912](https://github.com/keras-team/keras-hub/pull/2912) | **Model Specs: Gemma Text & FunctionGemma** | ~320 | Gemma text specs and FunctionGemma metadata overrides |
| **9** | [#2913](https://github.com/keras-team/keras-hub/pull/2913) | **Model Specs: Gemma Vision & Audio Metadata** | ~340 | Gemma 3n and Gemma 4 specs with audio and vision metadata |
| **10** | [#2895](https://github.com/keras-team/keras-hub/pull/2895) | **Model Specs: PaliGemma & Complete Registry** | ~320 | PaliGemma spec, vision encoder helpers, and complete spec registry |
| **11** | [#2914](https://github.com/keras-team/keras-hub/pull/2914) | **Text Export: PyTorch nn.Module Adapters** | ~270 | PyTorch nn.Module adapters, CPU default device scoping, and dtype helpers |
| **12** | [#2915](https://github.com/keras-team/keras-hub/pull/2915) | **Text Export: Tracing Pipeline & ExportPlan** | ~280 | ExportPlan, argument validation, sample input builders, and tracing engine |
| **13** | [#2916](https://github.com/keras-team/keras-hub/pull/2916) | **Text Export: Bundle Assembly & LlmMetadata Proto** | ~250 | SentencePiece extraction, LlmMetadata proto, and .litertlm bundle assembly |
| **14** | [#2917](https://github.com/keras-team/keras-hub/pull/2917) | **Text Export: Causal LM Hook & Driver** | ~310 | export_to_litertlm driver and CausalLM.export(format='litertlm') hook |
| **15** | [#2896](https://github.com/keras-team/keras-hub/pull/2896) | **Text Export: Complete Unit Test Suite** | ~370 | Complete unit test suite for text-only causal LM export |
| **16** | [#2918](https://github.com/keras-team/keras-hub/pull/2918) | **Tokenizer: HuggingFace Converter & Vocab Checks** | ~210 | BytePairTokenizer to HuggingFace tokenizer.json converter and vocab checks |
| **17** | [#2897](https://github.com/keras-team/keras-hub/pull/2897) | **Tokenizer: HuggingFace Integration & Bundling** | ~245 | hf_tokenizer_path integration, HF bundling, and end-to-end tokenizer tests |
| **18** | [#2919](https://github.com/keras-team/keras-hub/pull/2919) | **Test Harness: Shared TestCase & Parity Verifier** | ~390 | Shared TestCase export and numeric parity verifier harness |
| **19** | [#2920](https://github.com/keras-team/keras-hub/pull/2920) | **Test Harness: Enable Export Tests (First 8 Models)** | ~97 | Enable shared export tests for first 8 causal LM model families |
| **20** | [#2899](https://github.com/keras-team/keras-hub/pull/2899) | **Test Harness: Enable Export Tests (Remaining 8 Models)** | ~95 | Enable shared export tests for remaining 8 causal LM model families |
| **21** | [#2901](https://github.com/keras-team/keras-hub/pull/2901) | **Unsupported Models Declaration** | ~110 | Declare models currently unsupported by LiteRT-LM export (#2901) |
| **22** | [#2921](https://github.com/keras-team/keras-hub/pull/2921) | **Multimodal: Vision Adapter Support & Embeddings** | ~270 | Vision adapter support and image embedding prep in KerasHubLiteRTAdapter |
| **23** | [#2922](https://github.com/keras-team/keras-hub/pull/2922) | **Multimodal: ExportPlan Modality Fields & Inputs** | ~260 | Multimodal ExportPlan modality fields and sample input builders |
| **24** | [#2923](https://github.com/keras-team/keras-hub/pull/2923) | **Multimodal: TestCase Parity Verifier Harness** | ~446 | TestCase._verify_litertlm_multimodal_numerics parity verifier harness |
| **25** | [#2898](https://github.com/keras-team/keras-hub/pull/2898) | **Multimodal: End-to-End Gemma 3 Export Tests** | ~140 | End-to-end multimodal export and numeric parity tests for Gemma 3 |
| **26** | [#2924](https://github.com/keras-team/keras-hub/pull/2924) | **Separate Vision Encoder: 4-Section Bundle Assembly** | ~295 | 4-section bundle assembly support for separate vision encoders |
| **27** | [#2925](https://github.com/keras-team/keras-hub/pull/2925) | **Separate Vision Encoder: Audio Modality Support** | ~316 | Audio modality support, sample builders, separate vision bundle tests |
| **28** | [#2900](https://github.com/keras-team/keras-hub/pull/2900) | **Separate Vision Encoder: Complete Family Tests** | ~302 | End-to-end multimodal export tests for Gemma 3n, Gemma 4, and PaliGemma |

---

### Stacked PR Feature Flow

```mermaid
graph TD
    subgraph FOUNDATIONS ["1 - 2: Foundations"]
        PR01["1 · JAX Backend Unforce (#2892)"] --> PR02["2 · Layer Traceability Fixes (#2893)"]
    end

    subgraph SUBTREE_1 ["3 - 5: Traceable Ops"]
        PR02 --> PR03a["3 · Core Scope & Basic Shims (#2908)"]
        PR03a --> PR03b["4 · Indexing & Slicing Shims (#2909)"]
        PR03b --> PR03c["5 · Attention Shim & Parity (#2894)"]
    end

    subgraph SUBTREE_2 ["6 - 10: Model Specs & Registry"]
        PR03c --> PR04a["6 · Spec Base & Registry (#2910)"]
        PR04a --> PR04b["7 · Qwen Family Specs (#2911)"]
        PR04b --> PR04c["8 · Gemma Text & Function Specs (#2912)"]
        PR04c --> PR04d["9 · Gemma Vision & Audio Specs (#2913)"]
        PR04d --> PR04e["10 · PaliGemma Spec & Registry (#2895)"]
    end

    subgraph SUBTREE_3 ["11 - 15: Text Export Core"]
        PR04e --> PR05a["11 · PyTorch Adapters (#2914)"]
        PR05a --> PR05b["12 · Tracing Pipeline (#2915)"]
        PR05b --> PR05c["13 · Bundle Assembly (#2916)"]
        PR05c --> PR05d["14 · Causal LM Hook (#2917)"]
        PR05d --> PR05e["15 · Complete Test Suite (#2896)"]
    end

    subgraph SUBTREE_4 ["16 - 17: HuggingFace Tokenizer"]
        PR05e --> PR06a["16 · Tokenizer Converter (#2918)"]
        PR06a --> PR06b["17 · Tokenizer Integration (#2897)"]
    end

    subgraph SUBTREE_5 ["18 - 21: Shared Test Harness & Unsupported Models"]
        PR06b --> PR07a["18 · Shared Test Harness (#2919)"]
        PR07a --> PR07b["19 · Export Tests: First 8 Models (#2920)"]
        PR07b --> PR07c["20 · Export Tests: Remaining 8 Models (#2899)"]
        PR07c --> PR08["21 · Unsupported Models (#2901)"]
    end

    subgraph SUBTREE_6 ["22 - 25: Multimodal Baked-In Vision"]
        PR08 --> PR09a["22 · Vision Adapter Support (#2921)"]
        PR09a --> PR09b["23 · Multimodal Plan & Inputs (#2922)"]
        PR09b --> PR09c["24 · Parity Verifier Harness (#2923)"]
        PR09c --> PR09d["25 · Gemma 3 Multimodal Tests (#2898)"]
    end

    subgraph SUBTREE_7 ["26 - 28: Separate Vision Encoder & Audio"]
        PR09d --> PR10a["26 · 4-Section Vision Bundle (#2924)"]
        PR10a --> PR10b["27 · Audio Support & Tests (#2925)"]
        PR10b --> PR10c["28 · Complete Family Tests (#2900)"]
    end
```

**Comment by dhantule** [comment]
Hi @pctablet505, 
Closing this issue as all the PRs linked in this issue are merged / closed.
